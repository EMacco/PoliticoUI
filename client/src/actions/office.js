import axios from 'axios';
import { GET_ERRORS, SET_GOVERNMENT_OFFICES } from '@actions/types';
import { extractCandidateInfoFromOffice, checkIfUserExpressedInterest } from '../utils/offices';

const fetchOfficeDetailsByID = async id => {
  return await axios.get(`/offices/${id}`);
};

const fetchAllOffices = async () => {
  return (await axios.get('/offices')).data.data;
};

const fetchOfficeCandidates = async () => {
  return (await axios.get('/offices/candidates')).data.data;
};

const fetchAllInterests = async () => {
  return (await axios.get('/offices/interests')).data.data;
};

const collateResult = async id => {
  return await axios.post(`/office/${id}/result`);
};

const expressInterest = (partyId, officeId) => async dispatch => {
  try {
    const response = await axios.post('/offices/interests', { partyId, officeId });
    window.location.reload();
  } catch (err) {
    dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
  }
};

const fetchGovernmentOffices = userId => async dispatch => {
  try {
    const officeData = await fetchAllOffices();
    const candidates = await fetchOfficeCandidates();
    const interests = await fetchAllInterests();
    const offices = [];
    let alreadyApplied = false;

    for (let ind = 0; ind < officeData.length; ind += 1) {
      const { isCandidate } = extractCandidateInfoFromOffice(
        candidates,
        officeData[ind].id,
        userId
      );
      const isInterested = checkIfUserExpressedInterest(interests, officeData[ind].id, userId);

      if (isCandidate || isInterested) {
        alreadyApplied = true;
        break;
      }
    }

    for (let ind = 0; ind < officeData.length; ind += 1) {
      let candidateStatus = '';
      let officeBtn = 'add-party-btn';
      const { count, isCandidate, electionDate } = extractCandidateInfoFromOffice(
        candidates,
        officeData[ind].id,
        userId
      );

      const isInterested = checkIfUserExpressedInterest(interests, officeData[ind].id, userId);

      if (isCandidate) {
        candidateStatus = 'Approved candidate';
        officeBtn = 'casted-vote';
      } else if (isInterested) {
        candidateStatus = 'Awaiting approval';
        officeBtn = 'pending-vote';
      } else {
        candidateStatus = 'Express Interest';
        officeBtn = 'add-party-btn';
      }
      let electionDateString = new Date(electionDate).toDateString();
      if (electionDate === 'Not set') {
        electionDateString = 'Not set';
      }

      const { id, logourl, name, type } = officeData[ind];

      const officeObject = {
        officeId: id,
        officeLogo: logourl,
        officeName: name,
        officeType: type,
        electionDate: electionDateString,
        count,
        candidateStatus,
        officeBtn,
        alreadyApplied
      };

      offices.push(officeObject);
      if (ind === officeData.length - 1) {
        dispatch({ type: SET_GOVERNMENT_OFFICES, payload: offices });
      }
    }
  } catch (err) {
    dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
  }
};

export { fetchOfficeDetailsByID, collateResult, fetchGovernmentOffices, expressInterest };
