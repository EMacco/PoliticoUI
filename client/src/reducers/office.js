import { SET_GOVERNMENT_OFFICES, CLEAR_OFFICES } from '../actions/types';

const initialState = {
  offices: []
};

const officeReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_GOVERNMENT_OFFICES:
      return { ...state, offices: action.payload };
    case CLEAR_OFFICES:
      return { ...state, offices: [] };
    default:
      return state;
  }
};

export default officeReducer;
