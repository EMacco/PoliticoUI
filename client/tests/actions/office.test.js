import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { fetchGovernmentOffices } from '@actions/office';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const store = mockStore({
  auth: {},
  dashboard: {},
  parties: {},
  offices: {}
});

describe('Offices actions', () => {
  it('should dispatch SET_GOVERNMENT_OFFICES on successful fetch', async () => {
    await store.dispatch(fetchGovernmentOffices(1));
    expect(store.getActions()).toMatchSnapshot();
  });
});
