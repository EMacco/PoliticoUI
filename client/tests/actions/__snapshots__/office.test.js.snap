// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offices actions should dispatch SET_GOVERNMENT_OFFICES on successful fetch 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "alreadyApplied": true,
        "candidateStatus": "Approved candidate",
        "count": 1,
        "electionDate": "Fri Mar 01 2019",
        "officeBtn": "casted-vote",
        "officeId": 1,
        "officeLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452411/mtuaqvskpzlwlu331g1e.png",
        "officeName": "office of the president",
        "officeType": "federal",
      },
      Object {
        "alreadyApplied": true,
        "candidateStatus": "Express Interest",
        "count": 0,
        "electionDate": "Not set",
        "officeBtn": "add-party-btn",
        "officeId": 2,
        "officeLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551466689/nph9hky2tdup5qs9af5j.png",
        "officeName": "office of the governor of lagos state",
        "officeType": "state",
      },
    ],
    "type": "SET_GOVERNMENT_OFFICES",
  },
]
`;
